### PR TITLE
Auth as (None, None) should default to None

### LINKED ISSUE #195: Support downloading from sites with optional authentication
It seems that if we provide `(None, None)` as auth for a site that supports optional authentication, such as Jenkins, we get an authorization required response.

> $ mozdownload --url=https://ci.mozilla.org/
  INFO | Downloading from: https://ci.mozilla.org/
Traceback (most recent call last):
  File "/Users/dhunt/.virtualenvs/mozdownload/bin/mozdownload", line 8, in <module>
    load_entry_point('mozdownload==1.10', 'console_scripts', 'mozdownload')()
  File "/Users/dhunt/workspace/mozdownload/mozdownload/scraper.py", line 1016, in cli
    build.download()
  File "/Users/dhunt/workspace/mozdownload/mozdownload/scraper.py", line 274, in download
    r.raise_for_status()
  File "/Users/dhunt/.virtualenvs/mozdownload/lib/python2.7/site-packages/requests-1.2.2-py2.7.egg/requests/models.py", line 689, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 401 Client Error: Authorization Required

We should provide None for the auth argument rather than a tuple with None values. We also need a test for this, which might require a new entry in mozqa.com.

### PR DESCRIPTION
**Pull request by Ethcelon**
@whimboo 
This is an attempt to fix #195 
 > We should provide None for the auth argument rather than a tuple with None values.

### CODE REVIEW

**Comment by Ethcelon** [comment]
@whimboo is this the right way to do it?

File: tests/base_scraper/test_base_scraper.py [hunk 1]
```diff
@@ -74,5 +74,39 @@ def test_notimplementedexceptions(self):
         self.assertRaises(mozdownload.NotImplementedError,
                           scraper.build_filename, 'invalid binary')
 
+    def test_authentication(self):
+        """testing authentication"""
+        username = 'mozilla'
+        password = 'mozilla'
+        basic_auth_url = 'http://mozqa.com/data/mozqa.com/http_auth/basic/'
+        no_auth_url = 'https://ci.mozilla.org/' #does not need auth
```
Comment (whimboo) [inline tests/base_scraper/test_base_scraper.py]: I would propose that you split this test method into two different ones. Means one for basic auth, and the other for optional basic auth. Right now code is a bit mixed and makes it harder to understand.

File: tests/base_scraper/test_base_scraper.py [hunk 2]
```diff
@@ -74,5 +74,39 @@ def test_notimplementedexceptions(self):
...
+
+        # test with invalid authentication
+        # check if exception is thrown
```
Comment (whimboo) [inline tests/base_scraper/test_base_scraper.py]: The second comment line you can remove. The call to `assertRaises` makes that clear.

File: tests/base_scraper/test_base_scraper.py [hunk 3]
```diff
@@ -74,5 +74,39 @@ def test_notimplementedexceptions(self):
...
+        scraper = mozdownload.DirectScraper(directory=self.temp_dir,
+                                            url = basic_auth_url,
+                                            version=None,
+                                            log_level='ERROR')
```
Comment (whimboo) [inline tests/base_scraper/test_base_scraper.py]: You can keep the basic settings for the log level in all the cases.

File: tests/base_scraper/test_base_scraper.py [hunk 4]
```diff
@@ -74,5 +74,39 @@ def test_notimplementedexceptions(self):
...
+        scraper = mozdownload.DirectScraper(directory=self.temp_dir,
+                                            url = basic_auth_url,
+                                            version=None,
+                                            log_level='ERROR',
+                                            username=username,
+                                            password=password)
+        scraper.download()
+        self.assertTrue(os.path.isfile(os.path.join(self.temp_dir,
+                                                    'mozqa.com')))
+        # test with a site that requires no authentication
```
Comment (whimboo) [inline tests/base_scraper/test_base_scraper.py]: nit: `requires optional authentication`

**Comment by whimboo** [comment]
@Ethcelon, generally that test looks wonderful functionality wise! There are only some nits to fix. I think it would be ready to merge with the next update. Thank you a lot!

**Comment by whimboo** [comment]
@Ethcelon do you have a chance to get this PR finalized? It would be great to have it in the next release.

**Comment by Ethcelon** [comment]
@whimboo I will finish up by tonight!

File: tests/base_scraper/test_base_scraper.py (hunk 1 above)
Comment (Ethcelon) [inline tests/base_scraper/test_base_scraper.py]: I've split into 2 test cases: test_authentication and test_optional_authentication

File: tests/base_scraper/test_base_scraper.py (hunk 2 above)
Comment (Ethcelon) [inline tests/base_scraper/test_base_scraper.py]: Done

**Comment by Ethcelon** [comment]
@whimboo I've done the changes you suggested! Could you review this?

File: tests/base_scraper/test_base_scraper.py [hunk 5]
```diff
@@ -74,5 +74,39 @@ def test_notimplementedexceptions(self):
...
+        """testing with basic authentication"""
+        username = 'mozilla'
+        password = 'mozilla'
+        basic_auth_url = 'http://mozqa.com/data/mozqa.com/http_auth/basic/'
+
+        # test with invalid authentication
+        scraper = mozdownload.DirectScraper(directory=self.temp_dir,
+                                            url = basic_auth_url,
+                                            version=None)
+        self.assertRaises(requests.exceptions.HTTPError,scraper.download)
```
Comment (whimboo) [inline tests/base_scraper/test_base_scraper.py]: nit: please add a blank after the comma.

File: tests/base_scraper/test_base_scraper.py [hunk 6]
```diff
@@ -74,5 +74,39 @@ def test_notimplementedexceptions(self):
...
+
+        # testing with valid authentication
+        scraper = mozdownload.DirectScraper(directory=self.temp_dir,
+                                            url = basic_auth_url,
+                                            version=None,
+                                            username=username,
+                                            password=password)
```
Comment (whimboo) [inline tests/base_scraper/test_base_scraper.py]: I would suggest you also add two more tests when only the username or the password has been given.

File: tests/base_scraper/test_base_scraper.py [hunk 7]
```diff
@@ -74,5 +74,39 @@ def test_notimplementedexceptions(self):
...
+        scraper.download()
+        self.assertTrue(os.path.isfile(os.path.join(self.temp_dir,
+                                                    'mozqa.com')))
+
+    def test_optional_authenticaiton(self):
```
Comment (whimboo) [inline tests/base_scraper/test_base_scraper.py]: nit: spelling mistake.

**Comment by whimboo** [comment]
This looks way better. So we only have some nits left. @Ethcelon if you could finish this up soon, we can include it in the next version of mozdownload which we want to release ASAP. Thanks.

File: tests/base_scraper/test_base_scraper.py [hunk 8]
```diff
@@ -74,5 +74,39 @@ def test_notimplementedexceptions(self):
...
+    def test_authentication(self):
+        """testing with basic authentication"""
+        username = 'mozilla'
+        password = 'mozilla'
+        basic_auth_url = 'http://mozqa.com/data/mozqa.com/http_auth/basic/'
+
+        # test with invalid authentication
+        scraper = mozdownload.DirectScraper(directory=self.temp_dir,
+                                            url = basic_auth_url,
+                                            version=None)
```
Comment (whimboo) [inline tests/base_scraper/test_base_scraper.py]: I was wrong here. Sorry. We should indeed include the log level and set it to ERROR. It's important for our test runs. Thanks.

File: tests/base_scraper/test_base_scraper.py (hunk 5 above)
Comment (Ethcelon) [inline tests/base_scraper/test_base_scraper.py]: I should be catching these better! Done :)

**Comment by Ethcelon** [comment]
@whimboo the current functionality is that if we provide only username or only password to a site where we need no authentication, we will try to set the auth as (username, password) where the value not provided to the scraper defaults to 'None'. We get a 401 auth required error. Is this the desirable functionality? It seems to me that suppose we provide only username or only password then auth should default to (None, None)

**Comment by whimboo** [comment]
I think getting this 401 error is pretty fine given that the credentials weren't correct. I don't know if there are changes to only specify a username or password. @davehunt what do you think?

**Comment by davehunt** [comment]
It might be nice to catch this and raise a ValueError, but that's probably a separate patch.

**Comment by whimboo** [comment]
Sounds like an idea. Thanks Dave! I will get a follow-up filed soon. For this PR I think we are ready. It looks fine to me. I will get the commits squashed and the PR merged in a moment.

**Comment by whimboo** [comment]
This PR has been merged as https://github.com/mozilla/mozdownload/commit/8cd593f5108f60c6061918ac3d9c197b3473a78c

Thank you for working on this fix  @Ethcelon!

**Pull request closed without merging**